Replace internal json serialization with SimpleJson

jasonmead opened the pull request.

jchannon commented:
This looks good! Had a quick play with a very very simple app on the fork and all seemed good.

@NancyFx/most-valued-minions @thecodejunkie @grumpydev I'd like to get this in ASAP so we can rebase CoreCLR branch with it, can you review?

phillip-haydon commented:
:shipit:

phillip-haydon commented:
Unless @grumpydev or @thecodejunkie have any objections. I'll merge this tonight. (giving them time to see my message)

thecodejunkie commented:
Yeah I do. I think we need to consider the fact that if we get Simple-Json in, then we need to make sure we also build in the following concepts https://github.com/NancyFx/Nancy/wiki/Extending-Serialization-with-Converters

I think these are used quite a bit and it also helps you customize your serialization process

jchannon commented:
From what I can see of the code it seems the concept of JavaScript converters still exist - https://github.com/jasonmead/Nancy/blob/SimpleJson/src/Nancy/Json/Simple/NancySerializationStrategy.cs#L23-L36

thecodejunkie commented on src/Nancy/Json/Converters/TimeSpanConverter.cs:
```diff
@@ -54,6 +54,9 @@ private int GetValue(IDictionary<string, object> dictionary, string key)
             }
 
+            if (value is long)
+                return Convert.ToInt32((long)value);
```
Wrap in curly-braces

thecodejunkie commented on src/Nancy/Json/JavaScriptPrimitiveConverter.cs:
```diff
@@ -7,7 +7,17 @@ public abstract class JavaScriptPrimitiveConverter
+        {
+            return Serialize(obj, null);
+        }
         public abstract object Serialize(object obj, JavaScriptSerializer serializer);
```
Missing blank line before abstract method declaration

thecodejunkie commented on src/Nancy/Json/Simple/NancySerializationStrategy.cs:
```diff
@@ -0,0 +1,217 @@
+        static readonly DateTime MinimumJavaScriptDate = new DateTime(100, 1, 1, 0, 0, 0, DateTimeKind.Utc);
+
+
+        public NancySerializationStrategy() : this(false) { }
```
Braces on their own lines

thecodejunkie commented on src/Nancy/Json/Simple/NancySerializationStrategy.cs:
```diff
@@ -0,0 +1,217 @@
+            {
+                this.RegisterConverters(javaScriptConverters);
+            }
+            if (javaScriptPrimitiveConverters != null)
```
Missing blank line

thecodejunkie commented on src/Nancy/Json/Simple/NancySerializationStrategy.cs:
```diff
@@ -0,0 +1,217 @@
+
+        protected override string MapClrMemberNameToJsonFieldName(string clrPropertyName)
+        {
+            return this.retainCasing ? base.MapClrMemberNameToJsonFieldName(clrPropertyName) :
```
Change to format

```c#
return this.retainCasing 
   ? base.MapClrMemberNameToJsonFieldName(clrPropertyName)
   : clrPropertyName.ToCamelCase();
```

thecodejunkie commented on src/Nancy/Json/Simple/NancySerializationStrategy.cs:
```diff
@@ -0,0 +1,217 @@
+    using System.Linq;
+    using Nancy.Extensions;
+
+    internal class NancySerializationStrategy : PocoJsonSerializerStrategy
```
Changes required to this class

- Change the class to `public`
- Public members needs XML documentation (including the class itself)
- Remove double blank lines

jchannon replied:
problem with this is SimpleJson is marked as internal via an ifdef, we can change that to public if needs be but as it stands there would be an issue as this would be public but its inherited type would be internal. What should be done?

phillip-haydon replied:
Make it all public and remove the ifdef.

jchannon replied:
Only issue with that, it might not be an issue but users could then call Nancy.Simple.SimpleJson.Deserialize from anywhere in their app

phillip-haydon replied:
Users calling Deserialize, is not our problem. And if they do, it may be, because they want consistency rather than introducing json.net to serialize/deserialize since the Nancy impl was marked internal making it useless to the outside world. 

my 2c

jchannon replied:
Made the changes in my PR https://github.com/jasonmead/Nancy/pull/1

thecodejunkie commented on src/Nancy/Json/Simple/NancySerializationStrategy.cs:
```diff
@@ -0,0 +1,217 @@
+        public override object DeserializeObject(object value, Type type)
+        {
+            if (type.IsEnum || (ReflectionUtils.IsNullableType(type) && Nullable.GetUnderlyingType(type).IsEnum))									
+                return value == null ? null : Enum.Parse(ReflectionUtils.IsNullableType(type) ? Nullable.GetUnderlyingType(type) : type, value.ToString(), true);
```
- Missing curly braces
- The double ternary operator usage is messy, please refactor

thecodejunkie commented on src/Nancy/Json/Simple/NancySerializationStrategy.cs:
```diff
@@ -0,0 +1,217 @@
+            {
+                return primitiveConverter.Deserialize(value, type);
+            }
+            var valueDictionary = value as IDictionary<string, object>;
```
Missing blank line

thecodejunkie commented on src/Nancy/Json/Simple/NancySerializationStrategy.cs:
```diff
@@ -0,0 +1,217 @@
+                {
+                    var genericType = type.GetGenericTypeDefinition();
+                    var genericTypeConverter = this.FindJavaScriptConverter(genericType);
+                    if (genericTypeConverter != null)
```
Missing blank line

thecodejunkie commented on src/Nancy/Json/Simple/NancySerializationStrategy.cs:
```diff
@@ -0,0 +1,217 @@
+                    {
+                        var values = new Dictionary<string, object>(StringComparer.OrdinalIgnoreCase);
+                        var genericArguments = type.GetGenericArguments();
+                        for (int i = 0; i < genericArguments.Length; i++)
```
Missing blank line

thecodejunkie replied:
Use `var` instead of `int`

thecodejunkie commented on src/Nancy/Json/Simple/NancySerializationStrategy.cs:
```diff
@@ -0,0 +1,217 @@
+                        var genericArguments = type.GetGenericArguments();
+                        for (int i = 0; i < genericArguments.Length; i++)
+                        {
+                            values.Add(valueDictionary.Keys.ElementAt(i), this.DeserializeObject(valueDictionary.Values.ElementAt(i), genericArguments[i]));
```
There is a lot going on in this line, please split it up so it is easier to read

thecodejunkie commented on src/Nancy/Json/Simple/NancySerializationStrategy.cs:
```diff
@@ -0,0 +1,217 @@
+                output = dynamicValue.Value;
+                return true;
+            }
+            var inputType = input.GetType();
```
Missing blank line

thecodejunkie commented on src/Nancy/Json/Simple/NancySerializationStrategy.cs:
```diff
@@ -0,0 +1,217 @@
+                return true;
+            }
+            var inputType = input.GetType();
+            if (this.TrySerializeJavaScriptConverter(input, out output, inputType)) return true;
```
- Split to multiple lines
- Missing curly-braces

thecodejunkie commented on src/Nancy/Json/Simple/NancySerializationStrategy.cs:
```diff
@@ -0,0 +1,217 @@
+            var inputType = input.GetType();
+            if (this.TrySerializeJavaScriptConverter(input, out output, inputType)) return true;
+
+            if (this.TrySerializePrimitiveConverter(input, ref output, inputType)) return true;
```
- Split to multiple lines
- Missing curly-braces

thecodejunkie commented on src/Nancy/Json/Simple/NancySerializationStrategy.cs:
```diff
@@ -0,0 +1,217 @@
+            if (this.TrySerializeJavaScriptConverter(input, out output, inputType)) return true;
+
+            if (this.TrySerializePrimitiveConverter(input, ref output, inputType)) return true;
+            return base.TrySerializeKnownTypes(input, out output);
```
Missing blank-line

thecodejunkie commented on src/Nancy/Json/Simple/NancySerializationStrategy.cs:
```diff
@@ -0,0 +1,217 @@
+                }
+
+                if (time < MinimumJavaScriptDate)
+                    time = MinimumJavaScriptDate;
```
Missing curly-braces

thecodejunkie commented on src/Nancy/Json/Simple/NancySerializationStrategy.cs:
```diff
@@ -0,0 +1,217 @@
+                {
+                    dateTime = new DateTime(dateTime.Ticks, DateTimeKind.Local);
+                }
+                output = dateTime.ToString("o", CultureInfo.InvariantCulture);
```
Missing blank line

thecodejunkie commented on src/Nancy/Json/Simple/NancySerializationStrategy.cs:
```diff
@@ -0,0 +1,217 @@
+                        localTzOffset = time - input;
+                        suffix = "-";
+                    }
+                    suffix += localTzOffset.ToString("hhmm");
```
Missing blank line

thecodejunkie commented on src/Nancy/Json/Simple/NancySerializationStrategy.cs:
```diff
@@ -0,0 +1,217 @@
+                output = primitiveConverter.Serialize(input);
+                return true;
+            }
+            return false;
```
Missing blank line

thecodejunkie commented on src/Nancy/Json/Simple/NancySerializationStrategy.cs:
```diff
@@ -0,0 +1,217 @@
+        {
+            var primitiveConverter =
+                this.primitiveConverters.FirstOrDefault(x => x.SupportedTypes.Any(st => st.IsAssignableFrom(inputType)));
+            return primitiveConverter;
```
Missing blank line

thecodejunkie commented on src/Nancy/Json/Simple/NancySerializationStrategy.cs:
```diff
@@ -0,0 +1,217 @@
+                output = result.ToDictionary(kvp => this.MapClrMemberNameToJsonFieldName(kvp.Key), kvp => kvp.Value);
+                return true;
+            }
+            return false;
```
Missing blank line

thecodejunkie commented on src/Nancy/Json/Simple/NancySerializationStrategy.cs:
```diff
@@ -0,0 +1,217 @@
+        private JavaScriptConverter FindJavaScriptConverter(Type inputType)
+        {
+            var converter = this.converters.FirstOrDefault(x => x.SupportedTypes.Any(st => st.IsAssignableFrom(inputType)));
+            return converter;
```
Missing blank line

thecodejunkie commented on test/Nancy.Testing.Tests/BrowserFixture.cs:
```diff
@@ -562,11 +561,11 @@ public void Should_add_cookie_to_the_request()
 
             //Then
-            result.Body.AsString().ShouldNotBeEmpty();
+                var resultBody = result.Body.AsString();
```
Messed up indentation on these 2 lines

thecodejunkie commented on test/Nancy.Tests/Unit/Json/SimpleJsonFixture.cs:
```diff
@@ -0,0 +1,31 @@
+
+    public class SimpleJsonFixture
+    {
+
```
Remove blank line

thecodejunkie commented on test/Nancy.Tests/Unit/Json/SimpleJsonFixture.cs:
```diff
@@ -0,0 +1,31 @@
+    {
+
+        [Fact]
+        public void String_dictionary_values_are_Json_serialized_as_strings()
```
Tests are missing Given/When/Then

thecodejunkie commented on test/Nancy.Tests/Unit/Json/SimpleJsonFixture.cs:
```diff
@@ -0,0 +1,31 @@
+        }
+
+        [Fact]
+        public void Integer_dictionary_values_are_Json_serialized_as_integers()
```
Tests are missing Given/When/Then

thecodejunkie commented:
There is `a lot` of logic going on in `NancySerializationStrategy` all of these needs some sort of test coverage. There are date/time related code etc and nothing to make sure it works across time zones and so on

thecodejunkie commented:
@jchannon @phillip-haydon you guys are going back to code-review school =)

thecodejunkie commented:
There's also some reflection stuff going on.. we need to make sure that's compatible with the CoreCLR .. things like using `GetTypeInfo()`

phillip-haydon commented:
Why do that when I can threaten to merge and get you to finally take a look ;)

thecodejunkie commented:
@jasonmead Please don't take the review comments the wrong way. You have done some fantastic work in this pull-request, however since this is replacing something, that is very essential to Nancy, we have to set the bar high.

thecodejunkie commented:
@phillip-haydon touché ;)

jchannon commented:
Left a PR for @jasonmead with the required syntax changes

thecodejunkie commented on src/Nancy/Json/JavaScriptSerializer.cs:
```diff
@@ -29,57 +29,40 @@
+        public JavaScriptSerializer(bool registerConverters, int maxJsonLength, int recursionLimit, bool retainCasing, bool iso8601DateFormat, IEnumerable<JavaScriptConverter> converters, IEnumerable<JavaScriptPrimitiveConverter> primitiveConverters)
         {
-            _typeResolver = resolver;
+            _serializerStrategy = new NancySerializationStrategy(retainCasing, registerConverters, iso8601DateFormat, converters, primitiveConverters);
```
Can we pass in an instance of the `JsonConfiguration` class here instead? @jchannon? Perhaps have an overload that takes the config object and calls this ctor? Not sure the greedy ctor would be used, this class won't be used outside anyway right?

thecodejunkie replied:
If we pass in the config object then all the instance variables can be removed in favor of getting the values out of it instead.. i.e all places that does `this.retainCasing` (and so on) could be swapped for `this.config.RetainCasing` (and so on)

thecodejunkie commented:
Anything in this PR that accesses properties on `Type` need to be refactored to use `Type.GetTypeInfo()` or it won't be compatible with the CoreCLR. A quick glance shows that it's accessing things like `IsEnum`, `FullName`, `IsGenericType` and so on. Please limit the number of calls to `GetTypeInfo` to a minimum and instead reuse the instance (at least scope it to a method)

thecodejunkie commented on src/Nancy/Json/Simple/NancySerializationStrategy.cs:
```diff
@@ -0,0 +1,217 @@
+
+        private readonly List<JavaScriptConverter> converters = new List<JavaScriptConverter>();
+        private readonly List<JavaScriptPrimitiveConverter> primitiveConverters = new List<JavaScriptPrimitiveConverter>();
+        internal static readonly long InitialJavaScriptDateTicks = new DateTime(1970, 1, 1, 0, 0, 0, DateTimeKind.Utc).Ticks;
```
Perhaps these two `DateTime` should be changed into `DateTimeOffset` ?

thecodejunkie commented on src/Nancy/Json/Simple/NancySerializationStrategy.cs:
```diff
@@ -0,0 +1,217 @@
+            {
+                DateTime time = input.ToUniversalTime();
+
+                string suffix = "";
```
`string` should be changed to `var`

thecodejunkie commented on src/Nancy/Json/Simple/NancySerializationStrategy.cs:
```diff
@@ -0,0 +1,217 @@
+                    time = MinimumJavaScriptDate;
+
+                var ticks = (time.Ticks - InitialJavaScriptDateTicks) / 10000;
+                output = "\\/Date(" + ticks + suffix + ")\\/";
```
Using `string.Concat` would be nicer

jasonmead commented:
Latest should address most if not all issues presented here. I had to do a force push because I was following the contribution guidelines and rebased with upstream master. (I was hesitant about doing so but wasn't sure how to proceed).

@thecodejunkie, to address your concerns about passing JsonConfiguration into NancySerializationStrategy, I was able to refactor a bit and get the constructor parameters down to only needing the retainCasing boolean. I did this by removing support for recursion and max json length as discussed in the issue thread.

I added more tests for NancySerializationStrategy and everything has test coverage.

Let me know if anything else is needed.

jchannon commented on src/Nancy/Json/JavaScriptSerializer.cs:
```diff
@@ -29,459 +29,49 @@
 
-        public JavaScriptSerializer(JavaScriptTypeResolver resolver)
-            : this(resolver, false, 102400, 100, false, true, null, null)
+        public JavaScriptSerializer(bool registerConverters, bool retainCasing, IEnumerable<JavaScriptConverter> converters, IEnumerable<JavaScriptPrimitiveConverter> primitiveConverters)
```
This could take a JsonConfiguration object instead of the 4 arguments

jasonmead commented:
@thecodejunkie No offense taken. You have your standards. I'm just happy to contribute. (Again) For future reference, are the code formatting standards documented anywhere?

khellang commented:
@jasonmead It's somewhat documented under [CONTRIBUTING.md](https://github.com/NancyFx/Nancy/blob/master/CONTRIBUTING.md#style-guidelines). Otherwise you'd have to use existing code as an example, though we have some bit that are nicked from the Mono codebase etc. that have pretty horrible (non-existent) standards.

thecodejunkie commented on src/Nancy/Json/JavaScriptSerializer.cs:
```diff
@@ -29,459 +29,49 @@
 
-        public JavaScriptSerializer(JavaScriptTypeResolver resolver)
-            : this(resolver, false, 102400, 100, false, true, null, null)
+        public JavaScriptSerializer(bool registerConverters, JsonConfiguration configuration)
```
It would make more sense if these parameters swapped order, and perhaps even an overload without the `registerConverters` parameter?

thecodejunkie commented on src/Nancy/Json/JavaScriptSerializer.cs:
```diff
@@ -29,459 +29,49 @@
 
         public JavaScriptSerializer()
-            : this(null, false, 102400, 100, false, true, null, null)
+            : this(false,JsonConfiguration.Default)
```
Missing space after `,`

thecodejunkie commented on src/Nancy/Json/JavaScriptSerializer.cs:
```diff
@@ -29,459 +29,49 @@
-        bool _retainCasing;
-        JavaScriptTypeResolver _typeResolver;
-        bool _iso8601DateFormat;
+        bool retainCasing;
```
Don't think we need this, we can just read it straight of the configuration object

thecodejunkie commented on src/Nancy/Json/JavaScriptSerializer.cs:
```diff
@@ -29,459 +29,49 @@
-            : this(null, false, 2097152, 100)
-        {
-        }
+        readonly NancySerializationStrategy serializerStrategy;
```
explicitly mark as `private`

thecodejunkie commented on src/Nancy/Json/JavaScriptSerializer.cs:
```diff
@@ -29,459 +29,49 @@
-        internal JavaScriptTypeResolver TypeResolver
-        {
-            get { return _typeResolver; }
+                this.RegisterConverters(configuration.Converters, configuration.PrimitiveConverters);
```
While we're poking at this class, we might as well fix these. Wrap in curly-braces

thecodejunkie commented on src/Nancy/Json/JavaScriptSerializer.cs:
```diff
@@ -499,9 +87,7 @@ public void RegisterConverters(IEnumerable<JavaScriptPrimitiveConverter> primiti
             if (primitiveConverters == null)
                 throw new ArgumentNullException("primitiveConverters");
```
Wrap in curly-braces

thecodejunkie replied:
Wrap in curly-braces

thecodejunkie commented on src/Nancy/Json/JavaScriptSerializer.cs:
```diff
@@ -489,9 +79,7 @@ public void RegisterConverters(IEnumerable<JavaScriptConverter> converters)
             if (converters == null)
                 throw new ArgumentNullException("converters");
```
Wrap in curly-braces

thecodejunkie commented on src/Nancy/Responses/DefaultJsonSerializer.cs:
```diff
@@ -13,11 +13,12 @@
+
         private readonly TraceConfiguration traceConfiguration;
 
+        private bool? retainCasing;
```
This variable is never used anymore

jchannon commented:
Sent a PR to @jasonmead about some of the comments left. Oddly some of those lines referenced are already fixed so not sure why Github thinks they are still there

thecodejunkie commented:
@jchannon :+1:

jchannon commented:
Looks like the `GetTypeInfo()` has been added.

@thecodejunkie you want to re-review?

khellang commented:
@jasonmead You're doing a fine bit of work here. Thanks for doing this :grin::+1:

psibernetic commented:
Just want to check that the de/serialization will continue adjusting properties to lowerCase and vice-versa?

jchannon commented:
you mean camelCase?

yes this is covered

On 1 March 2016 at 21:40, Ovan Crone <notifications@github.com> wrote:

> Just want to check that the de/serialization will continue adjusting
> properties to lowerCase and vice-versa?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/NancyFx/Nancy/pull/2313#issuecomment-190917578>.
>

psibernetic commented:
thanks @jchannon

thecodejunkie commented on src/Nancy/Json/JavaScriptSerializer.cs:
```diff
@@ -26,536 +26,137 @@
-        bool _retainCasing;
-        JavaScriptTypeResolver _typeResolver;
-        bool _iso8601DateFormat;
+        internal static readonly JavaScriptSerializer DefaultSerializer = new JavaScriptSerializer(JsonConfiguration.Default);
```
Swap everything `internal` to `private`

thecodejunkie commented on src/Nancy/Json/JavaScriptSerializer.cs:
```diff
@@ -26,536 +26,137 @@
+        /// </summary>
+        /// <param name="input">JSON representation</param>
+        /// <typeparam name="T">The <see cref="Type"/> to deserialize into</typeparam>
+        /// <returns></returns>
```
Empty `<return>` XML comment

thecodejunkie commented on src/Nancy/Json/JavaScriptSerializer.cs:
```diff
@@ -26,536 +26,137 @@
+        /// Deserialize JSON
+        /// </summary>
+        /// <param name="input">JSON representation</param>
+        /// <returns></returns>
```
Empty `<returns>` XML comment

thecodejunkie commented on src/Nancy/Json/JavaScriptSerializer.cs:
```diff
@@ -26,536 +26,137 @@
         {
             if (converters != null)
+            {
                 RegisterConverters(converters);
```
Add `this` qualifier

thecodejunkie commented on src/Nancy/Json/JavaScriptSerializer.cs:
```diff
@@ -26,536 +26,137 @@
 
             if (primitiveConverters != null)
+            {
                 RegisterConverters(primitiveConverters);
```
Add `this` qualifier

thecodejunkie commented on src/Nancy/Json/JavaScriptSerializer.cs:
```diff
@@ -26,536 +26,137 @@
+        /// Register custom JSON converters
+        /// </summary>
+        /// <param name="primitiveConverters">An array of <see cref="JavaScriptPrimitiveConverter"/></param>
+        /// <exception cref="ArgumentNullException"></exception>
```
Empty XML comment

thecodejunkie commented on src/Nancy/Json/JavaScriptSerializer.cs:
```diff
@@ -26,536 +26,137 @@
+        /// Register custom JSON converters
+        /// </summary>
+        /// <param name="converters">An array of <see cref="JavaScriptConverter"/> to register</param>
+        /// <exception cref="ArgumentNullException"></exception>
```
Empty XML comment

thecodejunkie commented on src/Nancy/Json/JavaScriptSerializer.cs:
```diff
@@ -26,536 +26,137 @@
+        /// Serialize an object to JSON
+        /// </summary>
+        /// <param name="obj">The object to serialize</param>
+        /// <returns></returns>
```
Empty `<returns>` XML comment

thecodejunkie commented on src/Nancy/Json/Simple/NancySerializationStrategy.cs:
```diff
@@ -0,0 +1,231 @@
+        /// <summary>
+        /// Initializes a new instance of the <see cref="NancySerializationStrategy"/> class.
+        /// </summary>
+        public NancySerializationStrategy() : this(false)
```
I think there needs to be a `<remark>` comment on this, informing that using the default ctor means that `retainCasing` will be treated as `false`.

thecodejunkie commented on src/Nancy/Json/Simple/NancySerializationStrategy.cs:
```diff
@@ -0,0 +1,231 @@
+        /// Formats a property name to a JSON field name
+        /// </summary>
+        /// <param name="clrPropertyName">The property name to format</param>
+        /// <returns></returns>
```
Empty `<returns>` XML comment

thecodejunkie commented on src/Nancy/Json/Simple/NancySerializationStrategy.cs:
```diff
@@ -0,0 +1,231 @@
+        /// </summary>
+        /// <param name="value">The object to deserialize</param>
+        /// <param name="type">The type of object to deserialize</param>
+        /// <returns></returns>
```
Empty `<returns>` XML comment

thecodejunkie commented on src/Nancy/Json/Simple/NancySerializationStrategy.cs:
```diff
@@ -0,0 +1,231 @@
+        /// <returns></returns>
+        public override object DeserializeObject(object value, Type type)
+        {
+            if (type.IsEnum || (ReflectionUtils.IsNullableType(type) && Nullable.GetUnderlyingType(type).IsEnum))
```
Both of these `IsEnum` calls needs to happen on `type.GetTypeInfo().IsEnum` or it will not work on CoreCLR

thecodejunkie commented on src/Nancy/Json/Simple/NancySerializationStrategy.cs:
```diff
@@ -0,0 +1,231 @@
+                return javascriptConverter.Deserialize(valueDictionary, type);
+            }
+
+            if (!type.IsGenericType)
```
Needs to be `type.GetTypeInfo().IsGenericType` or it won't work on the CoreCLR

thecodejunkie commented on src/Nancy/Json/Simple/NancySerializationStrategy.cs:
```diff
@@ -0,0 +1,231 @@
+                return base.DeserializeObject(value, type);
+            }
+
+            var genericType = type.GetGenericTypeDefinition();
```
Needs to be `type.GetTypeInfo(). GetGenericTypeDefinition()` or it won't work on the CoreCLR

thecodejunkie commented on src/Nancy/Json/Simple/NancySerializationStrategy.cs:
```diff
@@ -0,0 +1,231 @@
+            }
+
+            var values = new Dictionary<string, object>(StringComparer.OrdinalIgnoreCase);
+            var genericArguments = type.GetGenericArguments();
```
Needs to be type.GetTypeInfo().GetGenericArguments() or it won't work on the CoreCLR

thecodejunkie commented on src/Nancy/Json/Simple/NancySerializationStrategy.cs:
```diff
@@ -0,0 +1,231 @@
+            var type = input as Type;
+            if (type != null)
+            {
+                output = type.GetTypeInfo().FullName;
```
This can be changed to `inputType.FullName` can't it?

jasonmead replied:
This case handles the instance where the input we wish to serialize is an instance of System.Type. Changing this to inputType.FullName causes the following test to fail: https://github.com/NancyFx/Nancy/blob/master/test/Nancy.Tests/Unit/JsonFormatterExtensionsFixtures.cs#L68

thecodejunkie replied:
Oh yeah, never mind :stuck_out_tongue_winking_eye:

thecodejunkie commented on src/Nancy/Json/Simple/NancySerializationStrategy.cs:
```diff
@@ -0,0 +1,231 @@
+            return true;
+        }
+
+        private JavaScriptPrimitiveConverter FindPrimitiveConverter(Type inputType)
```
What do you guys think about adding caching on this method and `FindJavaScriptConverter`? That would avoid us doing the same kind of LINQ query each a Type is serialized. 

We could just use `ConcurrentDictionary<Type, JavaScriptPrimitiveConverter>` and use the `GetOrAdd` pattern?

jasonmead replied:
Added in c1285b6

thecodejunkie commented:
Right, added what I think is the last round of review comments. This is very close on being pulled in now! :heart:

thecodejunkie commented:
@jasonmead :heart: :heart: :heart: Will take a closer look tomorrow morning, with a fresh pair of eyes, but I think this is ready to be pulled in!

prabirshrestha commented:
What are your plans on merging upstream with official SimpleJson repo?
These days I don't work on .net so I don't have time to look at CRs. But if anyone is interested I could start merging CRs if you guys approve it and in the long term actually give push permission.

jchannon commented:
:heart: :heart: :heart: :+1: :+1: :+1: Huge thanks @jasonmead . Nice work!

thecodejunkie commented:
@prabirshrestha Not sure we've made any changes to Simple-Json? I think @jchannon and @jasonmead can bring us up to speed

We've sent on PR to make it compatible with the CoreCLR https://github.com/facebook-csharp-sdk/simple-json/pull/70 by @yantrio

I think we'll just have to wait and see if, and if so how many, changes that will happen in Simple-Json inside of Nancy.. too early to say, we need to get 2.0 out so we shake the tree a bit

Yantrio commented:
The issue with that PR is what @khellang said, it's hard to target all dotnet5x monikers and we can't see a nice solution going upstream. I'm open to suggestions @prabirshrestha :)

jasonmead commented:
@prabirshrestha The only change that was made to SimpleJson was to make the reflection caches in PocoSerializationStrategy static so that we don't have to repopulate them every time we new up a NancySerializationStrategy. I can open a PR is you think people would find that useful.
